rust/src/rule_parser/xbits.rs

```diff
@@ -0,0 +1,323 @@
+        println!("tracker parsing issue");
+        return Err(());
+    }
+    if tracker[0].ne("track") {
+        SCLogError!("xbits track keyword parsing failed");
+        return Err(());
+    }
+    let tracker: u8 = match get_xbit_type(tracker[1]) {
+        Ok(val) => val,
+        Err(e) => {
+            SCLogError!("xbits tracker parsing failed: {:?}", e);
+            return Err(());
+        }
+    };
+    println!("tracker: {}", tracker);
+    let vartype = if tracker == DETECT_XBITS_TRACK_IPPAIR {
+        if bit_type == BIT_TYPE_HOSTBIT {
+            return Err(());
+        }
+        VAR_TYPE_IPPAIR_BIT
+    } else {
+        VAR_TYPE_HOST_BIT
+    };
+    println!("vartype: {}", vartype);
+    if args.len() == 1 {
+        return Ok((tracker, DETECT_XBITS_EXPIRE_DEFAULT, vartype));
+    }
+    let expire: Vec<&str> = args[1].trim().split(' ').map(|s| s.trim()).collect();
+    if expire.len() != 2 {
+        return Err(());
+    }
+    if expire[0].ne("expire") {
+        SCLogError!("xbits expire keyword parsing failed");
+        return Err(());
+    }
+    let expire: u32 = match expire[1].parse() {
+        Ok(val) => val,
+        Err(_) => {
+            SCLogError!("xbits expire parsing failed");
+            return Err(());
+        }
+    };
+    println!("expire: {}", expire);
+    if expire == 0 {
+        return Err(());
+    }
+    Ok((tracker, expire, vartype))
+}
+
+fn parse_xbits(arg: &str, bit_type: u8) -> Result<DetectBitsData, ()> {
+    let split_args: Vec<&str> = arg.trim().split(',').collect();
+    let res;
+    let mut cmd_name;
+
+    match split_args.len() {
+        1 => match split_args[0] {
+            "noalert" => {
+                return Ok(DetectBitsData {
+                    idx: 0,  // TODO is this a right thing to do?
+                    cmd: get_bit_cmd("noalert")?,
+                    name: std::ptr::null_mut(),
+                    tracker: 0,
+                    expire: 0,
+                    vartype: 0,
+                    fb_names: vec![].as_mut_ptr(),
+                    or_list: vec![].as_mut_ptr(),
+                    or_list_size: 0,
+                })
+            }
+            _ => {
+                SCLogError!("{:?} noalert parsing failed", get_bit_type_str(bit_type));
+                return Err(());
+            }
+        },
+        2 => {
+            if bit_type != BIT_TYPE_FLOWBIT {
+                SCLogError!("Erroneous number of arguments for {:?}", get_bit_type_str(bit_type));
+            }
+            cmd_name = match parse_cmd_name(split_args.clone(), bit_type) {
+                Ok(val) => val,
+                Err(_) => return Err(()),
+            };
+            return Ok(DetectBitsData {
+                idx: 0,
+                cmd: cmd_name.0,
+                name: (cmd_name.1).into_raw(),
+                tracker: 0,
+                expire: 0,
+                vartype: VAR_TYPE_FLOW_BIT,
+                fb_names: (cmd_name.2).as_mut_ptr(),
```

Comment:
Here, `cmd_name.2` is a vector of raw C strings and holds all the pointers to strings that it is supposed to.
Although, after this stage, on the C side, a pointer is received but it seems to not point to anything at all.
It was expected to be a `*mut *mut c_char`.
I'm not sure how should I try to traverse or reown the strings to read inside C so that may be very wrong but right now, `fb_names` seems to be dropped much earlier than when it should. I tried to `forget` about the vector pointer but could not make it work properly.